onnxruntime/core/framework/graph_partitioner.cc

```diff
@@ -533,6 +557,78 @@ static Status InlineNodes(Graph& graph, bool& modified_graph) {
   return Status::OK();
 }
 
+static Status IntelligentAotFunctionInliningImpl(const ExecutionProviders& execution_providers,
+                                                 const KernelRegistryManager& kernel_registry_mgr,
+                                                 Graph& graph,
+                                                 size_t& inlined_round_count) {
+  // handle testing edge case where optimizers or constant lifting results in graph with no nodes.
+  // doing it here saves all providers checking for this in GetCapability
+  if (graph.NumberOfNodes() == 0) {
+    return Status::OK();
+  }
+
+  for (auto& node : graph.Nodes()) {
+    for (auto& entry : node.GetAttributeNameToMutableSubgraphMap()) {
+      Graph* subgraph = entry.second;
+      // we pass through the FuncManager from the top level graph
+      ORT_RETURN_IF_ERROR(IntelligentAotFunctionInliningImpl(execution_providers,
+                                                             kernel_registry_mgr,
+                                                             *subgraph,
+                                                             inlined_round_count));
+    }
+  }
+
+  // Gather the candidates
+  InlinedVector<NodeIndex> inline_candidates;
+  for (auto& node : graph.Nodes()) {
+    if (node.CanBeInlined()) {
+      inline_candidates.push_back(node.Index());
+    }
+  }
+
+  if (inline_candidates.empty()) {
+    return Status::OK();
+  }
+
+  // Find out all the nodes that are already taken
+  InlinedHashSet<NodeIndex> claimed_by_ep;
+  for (const auto& ep : execution_providers) {
+    std::vector<std::unique_ptr<ComputeCapability>> capabilities;
+    ORT_RETURN_IF_ERROR(GetCapabilityForEPForAotInlining(graph, kernel_registry_mgr, *ep, capabilities));
+    for (auto& capability : capabilities) {
+      const auto& nodes = capability->sub_graph->nodes;
+      if (nodes.size() == 1) {
+        // Single node capability.
+        ORT_IGNORE_RETURN_VALUE(claimed_by_ep.insert(nodes[0]));
+      } else {
+        auto unclaimed_count = std::count_if(nodes.cbegin(), nodes.cend(), [&claimed_by_ep](NodeIndex node_index) {
+          return claimed_by_ep.count(node_index) == 0;
+        });
+
+        if (nodes.size() == narrow<size_t>(unclaimed_count)) {
+          claimed_by_ep.insert(nodes.cbegin(), nodes.cend());
+        }
+      }
+    }
+  }
+
+  // XXX: Insert version check. We need to collect all the versions
+  // that imported by the model. If the version is not supported by
+  // the model, we can not inline it.
+
+  for (auto node_index : inline_candidates) {
+    if (claimed_by_ep.count(node_index) == 0) {
+      auto* node = graph.GetNode(node_index);
+      if (node != nullptr) {
+        ORT_RETURN_IF_ERROR(graph.InlineFunction(*node));
+        ++inlined_round_count;
```

Comment:
nit: why is there a "round" in this variable name? Unlike the regular partitioning, there are no rounds in the inlining algorithm, right? Just wanted to make sure I didn't miss something.